v3-roadmap-and-design.md

```diff
@@ -0,0 +1,425 @@
+
+### Group API
+
+The main question is how closely we should follow the existing Zarr-Python implementation / `MutableMapping` interface. The table below shows the primary `Group` methods in Zarr-Python 2 and attempts to identify if and how they would be implemented in 3.0.
+
+| V2 Group Methods | `AsyncGroup` | `Group` | `h5py_compat.Group`` |
+| --- | --- | --- | --- |
+| `__len__` | `length` | `__len__` | `__len__` |
+| `__iter__` | `__aiter__` | `__iter__` | `__iter__` |
+| `__contains__` | `contains` | `__contains__` | `__contains__` |
+| `__getitem__` | `getitem` | `__getitem__` | `__getitem__` |
+| `__enter__` | N/A | N/A | `__enter__` |
+| `__exit__` | N/A | N/A | `__exit__` |
+| `group_keys` | `group_keys` | `group_keys` | N/A |
+| `groups` | `groups` | `groups` | N/A |
+| `array_keys` | `array_key` | `array_keys` | N/A |
+| `arrays` | `arrays`* | `arrays` | N/A |
+| `visit` | ? | ? | `visit` |
+| `visitkeys` | ? | ? | ? |
+| `visitvalues` | ? | ? | ? |
+| `visititems` | ? | ? | `visititems` |
+| `tree` | `tree` | `tree` | `Both` |
+| `create_group` | `create_group` | `create_group` | `create_group` |
+| `require_group` | N/A | N/A | `require_group` |
+| `create_groups` | ? | ? | N/A |
+| `require_groups` | ? | ? | ? |
+| `create_dataset` | N/A | N/A | `create_dataset` |
+| `require_dataset` | N/A | N/A | `require_dataset` |
+| `create` | `create_array` | `create_array` | N/A |
+| `empty` | `empty` | `empty` | N/A |
+| `zeros` | `zeros` | `zeros` | N/A |
+| `ones` | `ones` | `ones` | N/A |
+| `full` | `full` | `full` | N/A |
+| `array` | `create_array` | `create_array` | N/A |
+| `empty_like` | `empty_like` | `empty_like` | N/A |
+| `zeros_like` | `zeros_like` | `zeros_like` | N/A |
+| `ones_like` | `ones_like` | `ones_like` | N/A |
+| `full_like` | `full_like` | `full_like` | N/A |
+| `move` | `move` | `move` | `move` |
+
+**`zarr.h5compat.Group`**
+
+Zarr-Python 2.* made an attempt to align its API with that of [h5py](https://docs.h5py.org/en/stable/index.html). With 3.0, we will relax this alignment in favor of providing an explicit compatibility module (`zarr.h5py_compat`). This module will expose the `Group` and `Dataset` APIs that map to Zarr-Python’s `Group` and `Array` objects. 
+
+### Creation API
+
+Zarr-Python 2.* bundles together the creation and serialization of Zarr objects. Zarr-Python 3.* will make it possible to create objects in memory separate from serializing them. This will specifically enable writing hierarchies of Zarr objects in a single batch step. For example:
+
+```python
+
+arr1 = Array(shape=(10, 10), path="foo/bar", dtype="i4", store=store)
+arr2 = Array(shape=(10, 10), path="foo/spam", dtype="f8", store=store)
+
+arr1.save()
+arr2.save()
+
+# or equivalently
+
+zarr.save_many([arr1 ,arr2])
```

Comment:
This is a big point! So here is the batching that I mentioned before. It should be clarified, what the state of `store` is before save/save_many is called: is it usable, does it have some ephemeral storage (including data?) ?

Reply:
Good questions @martindurant - I've left a note that indicates this part of the doc is in need of some additional design effort.